server/src/main/java/org/apache/cassandra/sidecar/livemigration/LiveMigrationTaskImpl.java

```diff
@@ -0,0 +1,204 @@
+import org.apache.cassandra.sidecar.common.response.LiveMigrationTaskResponse;
+import org.apache.cassandra.sidecar.concurrent.ExecutorPools;
+import org.apache.cassandra.sidecar.config.LiveMigrationConfiguration;
+import org.apache.cassandra.sidecar.utils.SidecarClientProvider;
+
+/**
+ * Implementation of live migration task that handles file downloading with retry logic.
+ * Manages the lifecycle of data copy operations from source to destination instances.
+ */
+public class LiveMigrationTaskImpl implements LiveMigrationTask
+{
+    final LiveMigrationDataCopyRequest request;
+    final Map<Integer, OperationStatus> statusMap = new TreeMap<>();
+    private final InstanceMetadata instanceMetadata;
+    private final String source;
+    private final int port;
+    private final Vertx vertx;
+    private final ExecutorPools executorPools;
+    private final SidecarClientProvider sidecarClientProvider;
+    private final LiveMigrationConfiguration liveMigrationConfiguration;
+
+    // Indicates overall status of the operation (succeeded or failed).
+    // Future returned by downloader changes on next iteration. Using a separate future to track overall operation.
+    Promise<OperationStatus> promise;
+
+    @VisibleForTesting
+    volatile LiveMigrationFileDownloader downloader;
+    private volatile boolean cancelled = false;
+
+    public LiveMigrationTaskImpl(Vertx vertx,
+                                 ExecutorPools executorPools,
+                                 SidecarClientProvider sidecarClientProvider,
+                                 LiveMigrationConfiguration liveMigrationConfiguration,
+                                 LiveMigrationDataCopyRequest request,
+                                 String source,
+                                 int port,
+                                 InstanceMetadata instanceMetadata)
+    {
+        this.vertx = vertx;
+        this.executorPools = executorPools;
+        this.sidecarClientProvider = sidecarClientProvider;
+        this.liveMigrationConfiguration = liveMigrationConfiguration;
+        this.request = request;
+        this.instanceMetadata = instanceMetadata;
+        this.source = source;
+        this.port = port;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String id()
+    {
+        return request.id;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isCompleted()
+    {
+        return cancelled || (promise != null && promise.future().isComplete());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void start()
+    {
+        promise = Promise.promise();
+        int iteration = 0;
+        handleRetry(startInternal(iteration), iteration, request.maxIterations, promise);
+    }
+
+    public Future<OperationStatus> startInternal(int iteration)
+    {
+        downloader = LiveMigrationFileDownloader.builder()
```

Comment:
What if the task is cancelled already? Do you want to check the state and avoid creating the downloader? Otherwise, it is possible to create downloader and start downloading when the task is already cancelled.